File: gnote.rs/src/commands/tags.rs
use crate::utils;
use clap::Args;
use glob::glob;
use notify::{watcher, DebouncedEvent, RecursiveMode, Watcher};
use regex::Regex;
use shellexpand::tilde;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::mpsc::channel;
use std::time::Duration;

/// Generate tags file
#[derive(Args, Debug)]
pub struct TagsContext {
	/// Note directory
	#[clap(long="note-dir", default_value_t=String::from(tilde("~/env/note")))]
	note_dir: String,

	/// Note extension
	#[clap(long="note-extension", default_value_t=String::from(".gnote"))]
	note_extension: String,

	/// Note marker string
	#[clap(long="note-marker", default_value_t=String::from("∗"))]
	note_marker: String,

	/// Output tags file
	#[clap(long, default_value_t=String::from(tilde("~/tags")))]
	output: String,

	/// Cache directory
	#[clap(long="cache-dir", default_value_t=String::from(tilde("~/.cache/gnote")))]
	cache_dir: String,

	/// Watch mode
	#[clap(long)]
	watch: bool,
}

pub struct Tags {
	note_dir: PathBuf,
	note_extension: String,
	note_marker: String,
	output: PathBuf,
	cache_dir: PathBuf,
	watch: bool,
	pattern: Regex,
}

impl Tags {
	pub fn new(context: &TagsContext) -> Self {
		Self {
			note_dir: PathBuf::from(&context.note_dir),
			note_extension: context.note_extension.to_string(),
			note_marker: context.note_marker.to_string(),
			output: PathBuf::from(&context.output),
			cache_dir: PathBuf::from(&context.cache_dir),
			watch: context.watch,
			pattern: create_pattern(&context.note_marker),
		}
	}

	pub fn execute(&mut self) {
		self.check_dirs();
		self.create_tags();

		if self.watch {
			self.start_watch();
		}
	}

	/// Create tags file
	fn create_tags(&self) {
		self.empty_cache_dir();

		let note_paths = self.list_notes();
		for note_path in note_paths {
			self.create_tags_in_cache(&note_path)
		}

		self.create_all_tags_from_cache();

		println!("Created {}", self.output.display());
	}

	/// Make sure note directory exists
	fn check_dirs(&mut self) {
		self.note_dir = self.note_dir.canonicalize().expect(&format!(
			"Note directory not found: {}",
			self.note_dir.display()
		));

		fs::create_dir_all(&self.cache_dir).expect(&format!(
			"Failed to create cache dir: {}",
			self.cache_dir.display()
		));

		self.cache_dir = self.cache_dir.canonicalize().unwrap();
	}

	/// Empty cache directory
	fn empty_cache_dir(&self) {
		utils::empty_dir(&self.cache_dir).expect(&format!(
			"Failed to empty cache dir: {}",
			self.cache_dir.display()
		));
	}

	fn list_notes(&self) -> Vec<PathBuf> {
		glob(&format!(
			"{}/**/*{}",
			self.note_dir.to_string_lossy(),
			self.note_extension
		))
		.unwrap()
		.map(|file| {
			file
				.unwrap()
				.strip_prefix(&self.note_dir)
				.unwrap()
				.to_path_buf()
		})
		.collect()
	}

	/// Create <cache>/a.gnote
	fn create_tags_in_cache(&self, note_path: &Path) {
		let tags_content = self.extract_tags_from_file(note_path);
		self.write_tags_to_cache(note_path, &tags_content);
	}

	/// Read <note>/a.gnote and returns tags content
	fn extract_tags_from_file(&self, note_path: &Path) -> String {
		let full_node_path = &self.note_dir.join(note_path);
		let content = fs::read_to_string(&full_node_path).expect(&format!(
			"Failed to read file: {}",
			full_node_path.display()
		));
		extract_tags_from_text(&content, &full_node_path, &self.note_marker, &self.pattern)
	}

	fn write_tags_to_cache(&self, note_path: &Path, tags_content: &str) {
		if tags_content.is_empty() {
			return;
		}
		let full_note_cache_path = self.cache_dir.join(note_path);
		utils::write_with_create_dir(&full_note_cache_path, tags_content).expect(&format!(
			"Failed to write tags to cache: {}",
			note_path.display()
		));
	}

	fn create_all_tags_from_cache(&self) {
		let paths = glob(&format!(
			"{}/**/*{}",
			self.cache_dir.to_string_lossy(),
			self.note_extension
		))
		.unwrap();
		let mut all_tags_content = String::new();
		for path in paths {
			let path = path.unwrap();
			let tags_content = fs::read_to_string(&path).unwrap();
			all_tags_content.push_str(&format!("\n{}", &tags_content));
		}
		let all_tags_content = sort_tags(&all_tags_content);
		let result = format!("!_TAG_FILE_SORTED\t1\n{}", &all_tags_content);
		utils::write_with_create_dir(&self.output, &result).expect(&format!(
			"Failed to write tags to file: {}",
			self.output.display()
		));
	}

	fn start_watch(&self) {
		println!("Watching {}", self.note_dir.display());
		let (tx, rx) = channel();
		let mut watcher = watcher(tx, Duration::from_secs(1)).unwrap();
		watcher
			.watch(&self.note_dir, RecursiveMode::Recursive)
			.unwrap();
		loop {
			match rx.recv() {
				Err(e) => println!("watch error: {:?}", e),
				Ok(event) => match event {
					DebouncedEvent::Write(path) | DebouncedEvent::Create(path) => self.watch_changed(&path),
					DebouncedEvent::NoticeRemove(path) => self.watch_removed(&path),
					DebouncedEvent::Rename(_, new) => {
						self.watch_changed(&new);
					}
					_ => {}
				},
			}
		}
	}

	fn watch_changed(&self, full_note_path: &Path) {
		if !(full_note_path.is_file() && self.is_note_extension(full_note_path)) {
			return;
		}
		let note_path = &self.relative_note_path(full_note_path);
		println!("Changed: {}", note_path.display());
		self.create_tags_in_cache(note_path);
		self.create_all_tags_from_cache();
	}

	fn watch_removed(&self, full_note_path: &Path) {
		if !self.is_note_extension(full_note_path) {
			return;
		}
		let note_path = &self.relative_note_path(full_note_path);
		println!("Removed: {}", note_path.display());
		self.remove_tags_in_cache(note_path);
	}

	fn is_note_extension(&self, path: &Path) -> bool {
		let extension = path.extension().and_then(|v| v.to_str()).unwrap_or("");
		extension == &self.note_extension[1..]
	}

	fn relative_note_path(&self, full_note_path: &Path) -> PathBuf {
		full_note_path
			.strip_prefix(&self.note_dir)
			.unwrap()
			.to_path_buf()
	}

	fn remove_tags_in_cache(&self, note_path: &Path) {
		let full_note_cache_path = self.cache_dir.join(note_path);
		fs::remove_file(&full_note_cache_path).ok();
	}
}

fn create_pattern(note_marker: &str) -> Regex {
	Regex::new(&format!(r"{0}([^\s]+){0}", regex::escape(&note_marker))).unwrap()
}

fn extract_tags_from_text(text: &str, path: &Path, note_marker: &str, pattern: &Regex) -> String {
	let ids: Vec<&str> = pattern
		.captures_iter(&text)
		.map(|cap| cap.get(1).unwrap().as_str())
		.collect();

	let tags_content = ids
		.iter()
		.map(|id| {
			let jump = format!("/{0}{1}{0}", note_marker, id);
			format!("{}\t{}\t{}", id, path.to_string_lossy(), jump)
		})
		.collect::<Vec<String>>()
		.join("\n");

	return tags_content;
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_extract_tags_from_text() {
		let result = extract_tags_from_text(
			"
				hello world
				foo *bar*
				link step
				*baz* and 
			",
			Path::new("/a.gnote"),
			"*",
			&create_pattern("*"),
		);
		assert_eq!(
			result,
			"\
				bar\t/a.gnote\t/*bar*\n\
				baz\t/a.gnote\t/*baz*\
			"
		)
	}
}

fn sort_tags(tags_content: &str) -> String {
	let mut lines: Vec<_> = tags_content.lines().collect();
	lines.sort();
	lines.join("\n")
}

File: gnote.go/src/commands/tags.go
package commands

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/gutenye/gnote/gnote.go/src/utils"
	"github.com/yargevad/filepathx"
)

type Tags struct {
	NoteDir string `default:"${noteDir}" help:"Note directory"`

	NoteExtension string `default:"${noteExtension}" help:"Note extension"`

	NoteMarker string `default:"${noteMarker}" help:"Note marker"`

	Output string `default:"${output}" help:"Output tags file"`

	CacheDir string `default:"${cacheDir}" help:"Cache directory"`

	Watch bool `help:"Watch mode"`

	pattern *regexp.Regexp `hidden:""`
}

func (t *Tags) Run() error {
	t.pattern = createPattern(t.NoteMarker)

	t.checkDirs()
	t.createTags()

	if t.Watch {
		t.startWatch()
	}

	return nil
}

// Make sure note directory exists
func (t *Tags) checkDirs() {
	if _, err := os.Stat(t.NoteDir); os.IsNotExist(err) {
		log.Fatalf("Note directory not found. %s\n", err)
	}

	if err := os.MkdirAll(t.CacheDir, 0755); err != nil {
		log.Fatalf("Failed to create cache dir. %s\n", err)
	}
}

// Create tags file
func (t *Tags) createTags() {
	t.emptyCacheDir()

	notePaths := t.listNotes()
	for _, notePath := range notePaths {
		t.createTagsInCache(notePath)
	}

	t.createAllTagsFromCache()

	fmt.Printf("Created %s\n", t.Output)
}

// Empty cache directory
func (t *Tags) emptyCacheDir() {
	err := utils.EmptyDir(t.CacheDir)
	if err != nil {
		log.Fatal("Failed to empty cache dir. %s\n", err)
	}
}

func (t *Tags) listNotes() []string {
	pattern := fmt.Sprintf("%s/**/*%s", t.NoteDir, t.NoteExtension)
	files, err := filepathx.Glob(pattern)
	if err != nil {
		log.Fatalf("Failed to list notes. %s\n", err)
	}
	notes := []string{}
	for _, v := range files {
		notePath := strings.TrimPrefix(v, t.NoteDir)
		notes = append(notes, notePath)
	}
	return notes
}

// Create <cache>/a.gnote
func (t *Tags) createTagsInCache(notePath string) {
	tagsContent := t.extractTagsFromFile(notePath)
	t.writeTagsToCache(notePath, tagsContent)
}

// Read <note>/a.gnote and returns tags content
func (t *Tags) extractTagsFromFile(notePath string) string {
	fullNotePath, err := filepath.Abs(filepath.Join(t.NoteDir, notePath))
	if err != nil {
		log.Fatalf("Failed to get absolute path. %s\n", err)
	}
	content, err := os.ReadFile(fullNotePath)
	if err != nil {
		log.Fatalf("Failed to read file. %s\n", err)
	}
	return extractTagsFromText(string(content), fullNotePath, t.NoteMarker, t.pattern)
}

func (t *Tags) writeTagsToCache(notePath string, tagsContent string) {
	if tagsContent == "" {
		return
	}
	fullNoteCachePath := filepath.Join(t.CacheDir, notePath)
	err := utils.WriteFileWithMkdirAll(fullNoteCachePath, []byte(tagsContent), 0644)
	if err != nil {
		log.Fatalf("Failed to write tags to cache. %s\n", err)
	}
}

func (t *Tags) createAllTagsFromCache() {
	paths, err := filepathx.Glob(fmt.Sprintf("%s/**/*%s", t.CacheDir, t.NoteExtension))
	if err != nil {
		log.Fatalf("Failed to list notes. %s\n", err)
	}
	allTagsContent := ""
	for _, path := range paths {
		tagsContent, err := os.ReadFile(path)
		if err != nil {
			log.Fatalf("Failed to read tags from cache. %s\n", err)
		}
		allTagsContent += "\n" + string(tagsContent)
	}
	allTagsContent = sortTags(allTagsContent)
	result := fmt.Sprintf("!_TAG_FILE_SORTED\t1\n%s", allTagsContent)
	err = utils.WriteFileWithMkdirAll(t.Output, []byte(result), 0644)
	if err != nil {
		log.Fatalf("Failed to write tags: %s\n", err)
	}
}

func (t *Tags) startWatch() {
	fmt.Printf("Watching %s\n", t.NoteDir)

	// 	let (tx, rx) = channel();
	// 	let mut watcher = watcher(tx, Duration::from_secs(1)).unwrap();
	// 	watcher
	// 		.watch(&self.note_dir, RecursiveMode::Recursive)
	// 		.unwrap();
	// 	loop {
	// 		match rx.recv() {
	// 			Err(e) => println!("watch error: {:?}", e),
	// 			Ok(event) => match event {
	// 				DebouncedEvent::Write(path) | DebouncedEvent::Create(path) => self.watch_changed(&path),
	// 				DebouncedEvent::NoticeRemove(path) => self.watch_removed(&path),
	// 				DebouncedEvent::Rename(_, new) => {
	// 					self.watch_changed(&new);
	// 				}
	// 				_ => {}
	// 			},
	// 		}
	// 	}
	// }

}

func extractTagsFromText(content string, fullNodePath string, noteMarker string, pattern *regexp.Regexp) string {
	ids := []string{}
	matches := pattern.FindAllStringSubmatch(content, -1)
	for _, match := range matches {
		ids = append(ids, match[1:]...)
	}
	tagLines := []string{}
	for _, id := range ids {
		jump := fmt.Sprintf("/%s%s%s", noteMarker, id, noteMarker)
		tagLine := fmt.Sprintf("%s\t%s\t%s", id, fullNodePath, jump)
		tagLines = append(tagLines, tagLine)
	}
	return strings.Join(tagLines, "\n")
}

func createPattern(noteMarker string) *regexp.Regexp {
	marker := regexp.QuoteMeta(noteMarker)
	return regexp.MustCompile(fmt.Sprintf(`%s([^\s]+)%s`, marker, marker))
}

func sortTags(tagsContent string) string {
	lines := strings.Split(tagsContent, "\n")
	sort.Strings(lines)
	return strings.Join(lines, "\n")
}
